#! /bin/sh -e 
## amd64-int3-fix.dpatch by <fschueler@gmx.net>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Description: Revert int3 handling
## DP: Patch author: Andi Kleen <ak@muc.de>
## DP: Upstream status: not yet committed to BK

. $(dirname $0)/DPATCH

@DPATCH@

--- kernel-source-2.6.11-2.6.11-orig/arch/x86_64/kernel/kprobes.c	2005-03-02 08:38:12.000000000 +0100
+++ kernel-source-2.6.11-2.6.11/arch/x86_64/kernel/kprobes.c	2005-03-05 10:55:17.306986104 +0100
@@ -307,6 +307,8 @@
 	struct die_args *args = (struct die_args *)data;
 	switch (val) {
 	case DIE_INT3:
+		if (args->regs->cs & 3)
+			return NOTIFY_DONE;
 		if (kprobe_handler(args->regs))
 			return NOTIFY_STOP;
 		break;
